_posts/2018-04-26-IAM-managed-SSH-users.md

```diff
@@ -0,0 +1,173 @@
+---
+layout: post
+title: Managing EC2 SSH users with IAM
+banner: /assets/posts/2018-04-26-iam-managed-ssh-users/iamec2.png
+author: lvandonkersgoed
+---
+
+Servers in the cloud are supposed to be immutable and stateless. But what if the users allowed on a system are dynamic? You need a management system for your users!
+
+At Sentia we've learned to treat our server instances as [cattle](http://cloudscaling.com/blog/cloud-computing/the-history-of-pets-vs-cattle/). They should be replaceable at any time, which is achieved through statelessness, and every server in an autoscaling group should be the same, which is achieved by using immutable sources (like an AMI) for every instance.
+
+![Cattle](../../../assets/posts/2018-04-26-iam-managed-ssh-users/cattle.jpg)
+
+
+Managing users the old way
+--------------------------
+Most environments we build and maintain are based on Linux. In general, these systems have a limited amount of users; because of security and operability reasons, we do not often allow our customers to login to instances.
+
+In the few cases where the customer needs and is allowed SSH access to his systems, user management is easy. In our AMI creation process, we define the username(s) the customer can use, add a user in our automated AMI creation procedure and include their public SSH key in the `authorized_keys` file.
+
+The AMI is launched (as cattle), and the user has access to them. In case we need to cycle the keys - about once a year - we build a new AMI and update the autoscaling group. Easy as pie.
+
+Dynamic users
+-------------
+Recently we've worked on a big data project for a financial institution. Without going into details, the situation required data scientists to test and modify scripts on an instance in an AWS environment. These scripts would execute large processes on an EMR cluster. Customer SSH access into this instance was a realistic and appropriate requirement.
+
+Security guidelines dictate that multiple users should have their own user accounts. This allows you to evict users without affecting others, track user activity and assign users specific folders, disks or databases. 
+
+With that, the requirements for the solution were determined. We were to create a system that:
+- Would allow us to create and remove users on the fly, on multiple instances simultaneously (think cattle)
+- Would allow us to do this without terminating or rebooting the instance (boot time of the instance is about 30 minutes)
```

Comment:
I would remove the 30 minutes remark, not a pretty number and it doesn't add anything to the requirements per se, I think.